megatron/model/gmlp.py

```diff
@@ -0,0 +1,108 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from megatron.model.fused_softmax import FusedScaleMaskSoftmax
+from megatron.model.activations import get_activation
+from megatron.model.norms import get_norm
+from megatron import mpu
+
+
+class TinyAttention(nn.Module):
+    def __init__(self, neox_args, d_attn, d_ff, mask_fn):
+        super().__init__()
+        self.proj_qkv = nn.Linear(d_ff * 2, 3 * d_attn)
+        self.scale = d_attn ** -0.5
+        self.seq_len = neox_args.seq_length
+        self.proj_ffn = nn.Linear(d_attn, d_ff)
+        self.softmax = FusedScaleMaskSoftmax(
+            input_in_fp16=neox_args.precision == "fp16",
+            upper_triang_mask_fusion=neox_args.scaled_upper_triang_masked_softmax_fusion,
+            general_mask_fusion=neox_args.scaled_masked_softmax_fusion,
+            mask_func=mask_fn,
+            softmax_in_fp32=neox_args.attention_softmax_in_fp32,
+            scale=None)
+
+    def forward(self, x, attention_mask):
+        q, k, v = torch.chunk(self.proj_qkv(x), 3, dim=-1)
+        w = torch.einsum("bnd,bmd->bnm", q, k).unsqueeze(1) * self.scale
+        a = self.softmax(w, mask=attention_mask[..., :w.size(-2), :w.size(-1)]).squeeze(1)
+        x = torch.einsum("bnm,bmd->bnd", a, v)
+        return self.proj_ffn(x)
+
+
+class SpatialGatingUnit(nn.Module):
+    def __init__(self, neox_args, d_ff, d_attn=None, causal=True, mask_fn=None):
+        super().__init__()
+        self.causal = causal  
+        norm, eps = get_norm(neox_args)
+        self.norm = norm(d_ff, eps=eps)
+        self.proj = nn.Linear(neox_args.seq_length, neox_args.seq_length)
+        self.use_attn = d_attn is not None
+        if self.use_attn:
+            assert mask_fn is not None
+            self.attn = TinyAttention(neox_args=neox_args, d_attn=d_attn, d_ff=d_ff, mask_fn=mask_fn)
+        nn.init.zeros_(self.proj.weight)
+        nn.init.constant_(self.proj.bias, 1.)
+
+    def forward(self, x, attention_mask):
+        x = x.transpose(0, 1) # [s, b, d] -> [b, s, d]
+        res, gate = x.chunk(2, dim=-1)  # split along dim
+        gate = self.norm(gate)
+        weight = self.proj.weight
+        if self.causal:
```

Comment:
i added a https://github.com/lucidrains/g-mlp-pytorch/blob/main/g_mlp_pytorch/g_mlp_pytorch.py#L89 so it wouldn't break if you give it a sequence length < max sequence length - not sure if you can use it or not